File: java_solutions/230.kth-smallest-element-in-a-bst.java
/*
 * @lc app=leetcode id=230 lang=java
 *
 * [230] Kth Smallest Element in a BST
 */

class Solution {
    public int kthSmallest(TreeNode root, int k) {
        return Optional.of(((Function<TreeNode, Stream<Integer>>[]) new Function[1]))
            .filter(f -> (f[0] = n -> n == null ? Stream.empty() : Stream.of(Stream.of(n.left).flatMap(f[0]), Stream.of(n.val), Stream.of(n.right).flatMap(f[0])).flatMap(s -> s)) != null)
            .map(f -> f[0].apply(root).skip(k - 1).findFirst().get())
            .get();
    }
}

File: kotlin_solutions/230.kth-smallest-element-in-a-bst.kt
/*
 * @lc app=leetcode id=230 lang=kotlin
 *
 * [230] Kth Smallest Element in a BST
 */

class Solution { fun kthSmallest(root: TreeNode?, k: Int): Int { return toList(root).sorted()[k - 1] } private fun toList(node: TreeNode?): List<Int> { return node?.let { listOf(it.`val`) + toList(it.left) + toList(it.right) } ?: emptyList() } }

